lib/internal/perf/observe.js

```diff
@@ -0,0 +1,351 @@
+      throw new ERR_MISSING_ARGS('options.entryTypes', 'options.type');
+
+    switch (this[kType]) {
+      case undefined:
+        if (entryTypes !== undefined) this[kType] = kTypeMultiple;
+        if (type !== undefined) this[kType] = kTypeSingle;
+        break;
+      case kTypeSingle:
+        if (entryTypes !== undefined)
+          throw new ERR_INVALID_ARG_VALUE('options.entryTypes', entryTypes);
+        break;
+      case kTypeMultiple:
+        if (type !== undefined)
+          throw new ERR_INVALID_ARG_VALUE('options.type', type);
+        break;
+    }
+
+    if (this[kType] === kTypeMultiple) {
+      if (!ArrayIsArray(entryTypes)) {
+        throw new ERR_INVALID_ARG_TYPE(
+          'options.entryTypes',
+          'string[]',
+          entryTypes);
+      }
+      maybeDecrementObserverCounts(this[kEntryTypes]);
+      this[kEntryTypes].clear();
+      for (let n = 0; n < entryTypes.length; n++) {
+        if (kSupportedEntryTypes.includes(entryTypes[n])) {
+          this[kEntryTypes].add(entryTypes[n]);
+          maybeIncrementObserverCount(entryTypes[n]);
+        }
+      }
+    } else {
+      if (!kSupportedEntryTypes.includes(type))
+        return;
+      this[kEntryTypes].add(type);
+      maybeIncrementObserverCount(type);
+    }
+
+    if (this[kEntryTypes].size)
+      kObservers.add(this);
+    else
+      this.disconnect();
+  }
+
+  disconnect() {
+    maybeDecrementObserverCounts(this[kEntryTypes]);
+    kObservers.delete(this);
+    kPending.delete(this);
+    this[kBuffer] = [];
+    this[kEntryTypes].clear();
+    this[kType] = undefined;
+  }
+
+  takeRecords() {
+    const list = this[kBuffer];
+    this[kBuffer] = [];
+    return new PerformanceObserverEntryList(list);
+  }
+
+  static get supportedEntryTypes() {
+    return kSupportedEntryTypes;
+  }
+
+  [kMaybeBuffer](entry) {
+    if (!this[kEntryTypes].has(entry.entryType))
+      return;
+    this[kBuffer].push(entry);
+    kPending.add(this);
+    if (kPending.size)
+      queuePending();
+  }
+
+  [kDispatch]() { this[kCallback](this.takeRecords(), this); }
+
+  [kInspect](depth, options) {
+    if (depth < 0) return this;
+
+    const opts = {
+      ...options,
+      depth: options.depth == null ? null : options.depth - 1
+    };
+
+    return `PerformanceObserver ${inspect({
+      connected: kObservers.has(this),
+      pending: kPending.has(this),
+      entryTypes: ArrayFrom(this[kEntryTypes]),
+      buffer: this[kBuffer],
+    }, opts)}`;
+  }
+}
+
+function enqueue(entry) {
+  if (!isPerformanceEntry(entry))
+    throw new ERR_INVALID_ARG_TYPE('entry', 'PerformanceEntry', entry);
+
+  for (const obs of kObservers) {
+    obs[kMaybeBuffer](entry);
+  }
+}
+
+function observerCallback(name, type, startTime, duration, details) {
+  const entry =
+    new InternalPerformanceEntry(
+      name,
+      type,
+      startTime,
+      duration,
+      details);
+
+  if (details !== undefined) {
```

Comment:
If `details` is `null`, `Object.keys` throws.
```suggestion
  if (details != null) {
```

Reply:
Details will not be null here. This is called from the c++ internals and it should always be passed as undefined if not given.